### PR TITLE
[5.3] Fix #16593 and fix bug introduced in #16607

### LINKED ISSUE #16593: Blade "or" syntax not working with associative arrays
- Laravel Version: 5.3.25
- PHP Version: 5.6.4
- Database Driver & Version:

### Description:
The blade "or" construct has been broken by https://github.com/laravel/framework/commit/e8a9ed155fd10ac7b3595421055d1577fee246a0, the regex does not match when using associative arrays.

### Steps To Reproduce:
Use `{{ $array['foo'] or 'bar' }}` in any view and get "undefined index 'foo'".

### Failing test:
```
$compiler = new Illuminate\View\Compilers\BladeCompiler($this->getFiles(), __DIR__);
$this->assertEquals('<?php echo isset($name[\'bar\']) ? $name[\'bar\'] : \'foo\'; ?>', $compiler->compileString('{!! $name[\'bar\'] or \'foo\' !!}'));
```

### PR DESCRIPTION
**Pull request by manan-jadhav**
The regex now handles array syntax, with/without 'or' nicely. 

I've added tests to verify that this fixes #16593 .

### CODE REVIEW

**Comment by fernandobandeira** [comment]
👏 That was quick.

I see that you are testing for both
```
$name["key or space"] or "foo"
$name["key"] or "foo"
```

Seems good.

**Comment by taylorotwell** [comment]
Can you explain this regex in plain English?

**Comment by vlakoff** [comment]
Are you sure the two greedy `.+` you have added aren't problematic?

**Comment by CurosMJ** [comment]
@vlakoff There are unit tests which cover all possibilities, and they pass, so I dont think they're going to cause any problems now. If there are more cases in your head, let me know.

**Comment by CurosMJ** [comment]
This is the regex : `^(\$.+(?:\[[\'"].+[\'"]\])|[^\[\]\'"]*)(?:\s+or\s+)(.+)$`

Breaking it into three parts : 

1. `(\$.+(?:\[[\'"].+[\'"]\])|[^\[\]\'"]*)` : This gets captured in group #1 `$1`. 
Match things that start with a `$` then followed by 0 or more characters (`.+`) then followed by either 
- `(?:\[[\'"].+[\'"]\])` : String that opens with `['` or `["`, then contains characters (using greedy here makes sure the regex doesnt return too early), and then `']` or '"]'. This matches the array access statements.
- `[^\[\]\'"]*` : String that is empty, OR, do not contain any of `[`, `"`, `'`. This matches anything except the array access statements.

2. `(?:\s+or\s+) ` : Matches ` or `, that is, `or` with whitespace characters around it.

3. `(.+)` : This gets captured in group #2 `$2`. Matches any characters after `or`.

**Comment by taylorotwell** [comment]
Going to hold off on this for now on 5.3 as it's fairly edge case and I don't want to break things again on a minor release.

**Comment by vlakoff** [comment]
Another case to think of, and test, would be dynamic variable names:

```php
$foo = new \StdClass;
$foo->bar = 'content';
$dynName = 'bar';

{{ $foo->{$dynName} }}
```
